convert2rhel/checks.py

```diff
@@ -16,13 +16,194 @@
 
-from convert2rhel.pkghandler import call_yum_cmd
 
 logger = logging.getLogger(__name__)
 
+KERNEL_REPO_RE = re.compile("^.+:(?P<version>.+).el.+$")
+KERNEL_REPO_VER_SPLIT_RE = re.compile("\W+")
+LINK_KMODS_RH_POLICY = "https://access.redhat.com/third-party-software-support"
+
+
+def _convert_to_int_or_zero(s):
+    try:
+        return int(s)
+    except ValueError:
+        return 0
+
+
+def _get_kmod_comparison_key(path):
+    """Create a comparison key from the kernel module abs path.
+
+    Converts /lib/modules/5.8.0-7642-generic/kernel/lib/a.ko.xz ->
+    kernel/lib/a.ko.xz
+
+    Why:
+        all kernel modules lives in /lib/modules/{some kernel release}
+        if we want to make sure that the kernel package is really presented
+        on RHEL, we need to compare the full path, but because kernel release
+        might be different, we compare the relative paths after kernel release.
+    """
+    return "/".join(path.split("/")[4:])
+
+
+def get_host_kmods():
+    try:
+        # TODO make it work with utils.run_subprocess (now fails)
+        kernel_version = check_output(["uname", "-r"]).decode().rstrip()
+        kmod_str = check_output(
+            ["find", "/lib/modules/" + kernel_version, "-name", "*.ko*"]
+        ).decode()
+        assert kmod_str
+    except (subprocess.CalledProcessError, AssertionError):
+        logger.critical("Can't get list of kernel modules.")
+    else:
+        return set(
+            _get_kmod_comparison_key(path)
+            for path in kmod_str.rstrip("\n").split()
+        )
+
+
+def _repos_version_key(pkg_name):
+    try:
+        rpm_version = KERNEL_REPO_RE.search(pkg_name).group("version")
+    except AttributeError:
+        logger.critical(
+            "Unexpected package:\n%s\n is a source of kernel modules.",
+            pkg_name,
+        )
+    else:
+        return tuple(
+            map(
+                _convert_to_int_or_zero,
+                KERNEL_REPO_VER_SPLIT_RE.split(rpm_version),
+            )
+        )
+
+
+def get_most_recent_unique_kernel_pkgs(pkgs):
+    """Return the most recent versions of all kernel packages.
+
+    When we scan kernel modules provided by kernel packages
+    it is expensive to check each kernel pkg. Since each new
+    kernel pkg do not deprecate kernel modules we only select
+    the most recent ones.
+
+    :type pkgs: Iterable[str]
+    """
+
+    pkgs_groups = itertools.groupby(
+        pkgs, lambda pkg_name: pkg_name.split(":")[0]
+    )
+    return (
+        max(distinct_kernel_pkgs[1], key=_repos_version_key)
+        for distinct_kernel_pkgs in pkgs_groups
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
+    )
+
+
+def get_rhel_supported_kmods(_refresh_system_info=False):
+    """Return set of target RHEL supported kernel modules."""
+    # allowing to rerun resolve_system_info (for the needs of tests)
+    if _refresh_system_info:
+        system_info.resolve_system_info()
```

Comment:
I'm not in favor of code that is present only for the purpose of unit tests. I'm sure there's a way to write the tests to not require this workaround.

Reply:
me either but I was unable to find the way to solve it differently. The reason is such things like https://github.com/oamg/convert2rhel/blob/main/convert2rhel/systeminfo.py#L225. So the class initialized during the modules loading and  I can't affect it, without reinitialization. 

I believe it will be better to try to resolve this in a separate fix.

Reply:
@ZhukovGreen why it cannot reinitialized before the call of the function? To be honest I don't see the problem with that. what is preventing possibility to set the `system_info` in the test before you call the function?

Reply:
DI or dependency inversion could be used here:
```python
def get_rhel_supported_kmods(system_info):
```

I.e. make it responsibility of the caller to provide `system_info`. Another benefit is that in testing you can supply the object without the need for mocking/monkeypatching.

Reply:
This is essentially dead code - code that can never be executed at run-time.
After talking to @shaded-enmity, I agreed on not blocking the PR on things like this that do not directly affect the functionality. Here's a high priority task to remove the dead code and resolve the unit tests in a different way: https://issues.redhat.com/browse/OAMG-4634.

Reply:
As a DI it will work as well, but will be inconsistent, as this will be the only place where we would do this (at least now). @pirat89 advice works nicely, thx. OK, now we don't have this hack and it is moved to the fixture